### PR TITLE
7143 refactor analysis and widget json attributes

### LINKED ISSUE #7143: Refactor Analysis and Widget json attributes
Currently both `Analysis` and `Widget` have `_json` methods for json convertions for columns of that type. They can probably be refactored to use Rails serializers.

### PR DESCRIPTION
**Pull request by juanignaciosl**
This closes #7143.

### CODE REVIEW

**Comment by juanignaciosl** [comment]
@javitonino CR this, please.

File: app/controllers/carto/api/widgets_controller.rb [hunk 1]
```diff
@@ -35,7 +35,7 @@ def update
         update_params = params.slice(:order, :type, :title)
         update_params[:source_id] = source_id_from_params if source_id_from_params
         @widget.update_attributes(update_params)
-        @widget.options = params[:options].to_json if params[:options]
+        @widget.options = params[:options] if params[:options]
```
Comment (javitonino) [inline app/controllers/carto/api/widgets_controller.rb]: I think you can remove `if params[:options]` here.

File: app/models/carto/analysis.rb [hunk 2]
```diff
@@ -46,6 +43,12 @@ def map
 
   private
 
+  def analysis_definition_must_be_json
+    unless analysis_definition.nil? || analysis_definition.is_a?(Hash)
```
Comment (javitonino) [inline app/models/carto/analysis.rb]: This is a bit confusing to read. I think this is easier to read:

`if analysis_definition && !analysis_definition.is_a?(Hash)`

But up to you.

File: lib/assets/javascripts/cdb [hunk 3]
```diff
@@ -1 +1 @@
-Subproject commit ea0290f51d64a2772c69af6ef67dfa7e9af7a9d0
+Subproject commit 4176c03f6e04488d6c3cb7647f37d40099a77831
```
Comment (javitonino) [inline lib/assets/javascripts/cdb]: Should you commit this? Just asking in case it slipped off.

File: spec/factories/analyses.rb [hunk 4]
```diff
@@ -5,15 +5,13 @@
...
-                                    }
-                                  }
-                                ).freeze
+  SOURCE_ANALYSIS_DEFINITION = {
+    "id": unique_string,
+    "type": "source",
+    "params": {
+      "query": "select * from subway_stops"
+    }
+  }.freeze
```
Comment (javitonino) [inline spec/factories/analyses.rb]: I'm pretty sure you can rewrite this without quotes on the keys (e.g: `type: "source"`). I think ruby does it automatically anyways:
```
irb(main):002:0> { "type": "source" }
=> {:type=>"source"}
```
AFAIK, if you want string keys, you need to use the `=>` operator.

File: spec/models/carto/analysis_spec.rb [hunk 5]
```diff
@@ -14,35 +14,35 @@
...
       Carto::Analysis.new(analysis_definition: nil).natural_id.should eq nil
-      Carto::Analysis.new(analysis_definition: '{}').natural_id.should eq nil
-      Carto::Analysis.new(analysis_definition: '{ "wadus": 1 }').natural_id.should eq nil
+      Carto::Analysis.new(analysis_definition: {}).natural_id.should eq nil
+      Carto::Analysis.new(analysis_definition: { "wadus": 1 }).natural_id.should eq nil
     end
 
     it 'returns id if analysis definition has id at the first level' do
-      Carto::Analysis.new(analysis_definition: '{ "id": "a1" }').natural_id.should eq 'a1'
+      Carto::Analysis.new(analysis_definition: { "id": "a1" }).natural_id.should eq 'a1'
```
Comment (javitonino) [inline spec/models/carto/analysis_spec.rb]: More string keys here (and the previous testcase).

File: app/models/carto/widget.rb [hunk 6]
```diff
@@ -49,9 +45,7 @@ def writable_by_user?(user)
   private
 
   def options_must_be_json
-    options_json
-  rescue JSON::ParserError
-    errors.add(:options, 'is wrongly formatted (invalid JSON)')
+    errors.add(:options, 'wrongly formatted (not a Hash or invalid JSON)') unless options.nil? || options.is_a?(Hash)
```
Comment (javitonino) [inline app/models/carto/widget.rb]: Same condition as above. If you decide to change it, do both.

Comment (javitonino) [inline app/models/carto/widget.rb]: Also, as we have the same code twice (and probably more in the future, as we add more json fields), maybe we could extract it to an [ActiveRecord validator](http://guides.rubyonrails.org/active_record_validations.html#custom-validators). Could also apply it to the fields that are already using `CartoJsonSerializer`.

File: spec/requests/carto/api/analyses_controller_spec.rb [hunk 7]
```diff
@@ -73,7 +73,7 @@ def verify_analysis_response_body(response_body, analysis)
         :source_analysis,
         visualization_id: @visualization.id,
         user_id: @user.id,
-        analysis_definition: %({"id": "#{UUIDTools::UUID.random_create}"})
+        analysis_definition: { "id": UUIDTools::UUID.random_create.to_s }
```
Comment (javitonino) [inline spec/requests/carto/api/analyses_controller_spec.rb]: And another string key, just marking them to make them easier to find.

**Comment by javitonino** [comment]
I really :heart: this. I left some comments, but nothing major. Take a look at them, but this is :package:-ready.

File: lib/assets/javascripts/cdb (hunk 3 above)
Comment (juanignaciosl) [inline lib/assets/javascripts/cdb]: No, not at all, I overlooked it when I reviewed the PR, thanks!

File: app/controllers/carto/api/widgets_controller.rb (hunk 1 above)
Comment (juanignaciosl) [inline app/controllers/carto/api/widgets_controller.rb]: ok

File: app/models/carto/analysis.rb (hunk 2 above)
Comment (juanignaciosl) [inline app/models/carto/analysis.rb]: ok

File: app/models/carto/widget.rb (hunk 6 above)
Comment (juanignaciosl) [inline app/models/carto/widget.rb]: Sure! Great suggestion!

File: spec/factories/analyses.rb (hunk 4 above)
Comment (juanignaciosl) [inline spec/factories/analyses.rb]: Yep, it's legacy from the `%(` version.

File: spec/models/carto/analysis_spec.rb (hunk 5 above)
Comment (juanignaciosl) [inline spec/models/carto/analysis_spec.rb]: ok

File: spec/requests/carto/api/analyses_controller_spec.rb (hunk 7 above)
Comment (juanignaciosl) [inline spec/requests/carto/api/analyses_controller_spec.rb]: ok

**Comment by juanignaciosl** [comment]
@javitonino detailed CR! You might want to take a look (after tests pass). I've taken into account your suggestions (first Rails custom validator for CartoDB, really simple! :-) ) and also fixed an English typo.

**Comment by javitonino** [comment]
:+1:

I'm also going to close https://github.com/CartoDB/cartodb/issues/6818 as the new serializer + validator do pretty much the same that gem does, but adapted to our needs.

**Comment by juanignaciosl** [comment]
Ok, thanks!

**Pull request merged**